tools/AutomationTestFramework/Management/RunbookProcessor.ps1

```diff
@@ -0,0 +1,207 @@
+            $content = switch -wildcard ($_) {
+                "*$loginParamsTemplate" {
+                    $_ -replace $loginParamsTemplate, "'$($template.AutomationConnectionName)' '$($template.SubscriptionName)'"
+                } $testListTemplate {
+                    Create-TestContent $folder.Path | Out-String
+                } default {
+                    $_
+                }
+            }
+            $content | Add-Content $bookPath
+        }
+        Write-Verbose "$bookPath generated."
+    }
+}
+
+# https://www.codeisahighway.com/how-to-easily-and-silently-obtain-accesstoken-bearer-from-an-existing-azure-powershell-session/
+function Get-AzureRmAccessToken() {
+    $azureRmProfile = [Microsoft.Azure.Commands.Common.Authentication.Abstractions.AzureRmProfileProvider]::Instance.Profile
+    if(-not $azureRmProfile.Accounts.Count) {
+        Write-Error "Ensure you have logged in before calling this function."
+        return
+    }
+  
+    $currentAzureContext = Get-AzureRmContext
+    $profileClient = New-Object Microsoft.Azure.Commands.ResourceManager.Common.RMProfileClient($azureRmProfile)
+    Write-Debug ("Getting access token for tenant" + $currentAzureContext.Subscription.TenantId)
+    $token = $profileClient.AcquireAccessToken($currentAzureContext.Subscription.TenantId)
+    $token.AccessToken
+}
+
+function Start-Runbooks ([hashtable] $automation, [string] $runbooksPath) {
+    # Remove any previous jobs from this PS session
+    Get-Job | Remove-Job
+
+    foreach ($runbook in Get-ChildItem $runbooksPath) {
+        $bookName = $runbook.BaseName
+        Write-Verbose "Uploading '$bookName' runbook..."
+        $null = Import-AzureRmAutomationRunbook -Path $runbook.FullName -Name $bookName -type PowerShell -AutomationAccountName $automation.AccountName -ResourceGroupName $automation.ResourceGroupName -LogVerbose $true -Force -ErrorAction Stop
+        Write-Verbose "Publishing '$bookName' runbook..."
+        $null = Publish-AzureRmAutomationRunbook -Name $bookName -AutomationAccountName $automation.AccountName -ResourceGroupName $automation.ResourceGroupName -ErrorAction Stop
+        
+        $accessToken = Get-AzureRmAccessToken
+        $accountId = $(Get-AzureRmContext).Account.Id
+        $subscriptionName = $(Get-AzureRmContext).Subscription.Name
+        Write-Verbose "Starting '$bookName' runbook..."
+        $null = Start-Job `
+            -Name $bookName `
+            -ArgumentList $accessToken,$accountId,$subscriptionName,$bookName,$automation `
+            -ScriptBlock {
+            param($accessToken,$accountId,$subscriptionName,$bookName,$automation)
+            Login-AzureRmAccount -AccessToken $accessToken -AccountId $accountId -Subscription $subscriptionName
+            Start-AzureRmAutomationRunbook -Name $bookName -AutomationAccountName $automation.AccountName -ResourceGroupName $automation.ResourceGroupName -ErrorAction Stop -Wait -MaxWaitSeconds 3600
+        }
```

Comment:
So here, you don't need to login inside the Start-Job call, instead, you can pass in the value of Get-AzureRMContext in the argument list in your script block, and use the ```-DefaultProfile``` parameter :

```
Start-Job {param($ctx) Start-AzureRMAutomationRunbook -AzureRmContext $ctx ...} -ArgumentList (Get-AzureRmContext)
```
 Note that, to be really slick, you can pass in the argument list as a hashtable, and use splatting, like this: https://github.com/Azure/azure-powershell/blob/experiments/experiments/Compute.Experiments/AzureRM.Compute.Experiments.psm1#L109

Reply:
This is one of those places where I got it to work and was happy with it. Initially, using Start-Job, it is completely non-obvious that it creates a new AppDomain. I was having login issues and was trying to figure out what was wrong. Then, I realized I needed to pass everything it. At first, I though it worked similar to a lambda expression, so I thought (at least) the current scope was passed in, but I was wrong.

But, the splatting suggestion is a good one. The idea of using the context instead of logging in is... not something that I ever saw suggested anywhere on the internet. The way Context is presented is that there is a bunch of 'behind the scenes' things it does when you log in, and the Context object is just more of an interface to some of the things that are set. So, passing the Context to individual AzureRM commands would have never crossed my mind. Is this another possible documentation thing?